### ISSUE TITLE
Bug: [prefer-optional-chain] invalid fix when the variable is read from an optional chain

### ISSUE THREAD

**Issue by OpportunityLiu** [typescript-eslint/typescript-eslint#11906]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgEJysg3gWAFDLIAeAXMiAK4C2ARtANz6ECeZltD%2BAvvvqJLEQoAYgHtR2JshoYy6KIwLS4ALzbU6C7rzwwKIBGGCiQyKnFAAKGOLJiJAH2T6AJhBigILgDTKocjABKMgA3UWAXSSUEEwBnMGRRABsXeWQAXmQbUQB%2BADoZLSVgGGRLZNSMZAAyasSU%2BTyiDPTMwqbAqMJCbILVDOQARkVCHjwuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

### Repro Code

```TypeScript
interface Bar {
  x: number;
  y: number;
}

interface Foo {
  bar: Bar;
  baz: number;
}

function main(foo: Foo | undefined, bar: Bar): void {
  const oldBar = foo?.bar;
  if (oldBar && oldBar.x === bar.x) {
    foo.baz = 1;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/prefer-optional-chain": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

No suggest

### Actual Result

A bad fix that leads to a TypeScript Error

### Additional Info

Seems the fix itself is correct, but TypeScript doesn't accept it. Maybe related #11840

**Comment by JoshuaKGoldberg** [comment]
> but TypeScript doesn't accept it

Yeah, this is an unfortunate limitation of the current TS _"variable depends on another variable"_ type narrowing systems.

**Comment by kirkwaiblinger** [comment]
To add to that, I think part of it is that in principle (ignoring types for a second) `bar.x` _could_ be `undefined`. In this case, the `oldBar?.x === bar.x` could be satisfied by `oldBar` being `null`/`undefined`, which could be due to `foo` being `null`/`undefined`.

Therefore, `oldBar?.x === bar.x` way doesn't manifestly guarantee at runtime that `oldBar` isn't `undefined` nearly as strongly as `oldBar && oldBar.x === bar.x`, which really does guarantee it, even without any type information. 

Even so, I'm quite surprised that TS gets the narrowing correct on the provided code at all. It doesn't even get it right with an explicit null check!

```ts
function main(foo: Foo | undefined, bar: Bar): void {
  const oldBar = foo?.bar;
  if (oldBar != null && oldBar.x === bar.x) {
    foo.baz = 1; // TS error: foo is possibly undefined
  }
}
```

So it seems like almost a (fortunate) fluke that TS accepts your code? 

---

@JoshuaKGoldberg What is the resolution that you'd be accepting PRs for here? I'd have thought this would be a wontfix edge case

**Comment by JoshuaKGoldberg** [comment]
> (ignoring types for a second) `bar.x` _could_ be `undefined`. In this case, ...

Sorry, I don't follow the train of logic here? I see the code snippets as benefiting from more specific logic because `bar.x` _couldn't_ be `undefined`.

Marking this as an edge case because TypeScript only understands some forms of `if (<truthiness for oldBar.x> === bar.x)` -> `foo.baz` being allowed feels ... off to me? That would be a TypeScript bug to file, not us, no?

**Comment by kirkwaiblinger** [comment]
@JoshuaKGoldberg 

Sorry, this is hairy, and difficult to summarize 🙃 

1. The rule's preference for converting `a && a.b === c` to `a?.b === c` takes the code to a state of lesser safety guarantee. Before, assigning `foo.baz = 1` is provable to be safe _regardless_ of the type system being correct at runtime; after, assigning `foo.baz = 1` is guaranteed to be safe _only if_ the type system is correct at runtime.

1. The proof of safety for both the before and after cases is not trivial. It is surprising to me that TS does it at all (hence my use of the word "fluke"), and TS is inconsistent about doing it. I do not think that we should hardcode assumptions about whether TS "correctly" narrows `foo` into the rule logic, nor attempt to reason about cross-variable narrowings ourselves.

1. Therefore, I concluded the reported issue would be an inherent edge case/limitation of the rule's `a && a.b === c` -> `a?.b === c` preference, not an instance of a pattern that we can define, detect, and exempt from the rule.

1. However, by the fact that you initially marked this as "Accepting PRs" I assume you did not arrive at the same conclusion in point 3, and so I'm trying to reconcile our perspectives. In this case, it seemed it would be simplest to work backwards from the conclusion: if we were to "solve" this with a PR, what would that "solution" do? If you have an idea in mind for that, it would obviously disprove my reasoning chain in Points 2/3; if not, I don't see what choice we have other than "wontfix".

**Comment by JoshuaKGoldberg** [comment]
Aha! Thank you, that's very helpful. I think I understand and the difference is primarily in step 2:

> do not think that we should hardcode assumptions about whether TS "correctly" narrows foo into the rule logic, nor attempt to reason about cross-variable narrowings ourselves

I think we're coming at that difference from different directions. My thought is to assume that when TypeScript generally tries to do something, that thing is meant to be wholly supported. Any gap in support should be treated as a transient bug in TypeScript. Concretely: given that TypeScript supports the `if (oldBar &&` form, `if (oldBar != null &&` should be considered _meant to be_ supported as well. 

From another direction: 

> converting `a && a.b === c` to `a?.b === c` ... is guaranteed to be safe _only if_ the type system is correct

We've generally made that _"if the type system says it's ok, it's ok"_ assumption most of the time in most rules. The `?.` & `??` rules a little less so because TypeScript has a lot more edge cases (and in e.g. this case IMO bugs) around them. But I think this case is common enough that we should be optimistic about TypeScript's support for it. At the end of the day, the type system told us this is ok - and it's a pretty useful thing for users when it works.

**Comment by kirkwaiblinger** [comment]
I get where you're coming from on this, but then my question is still: what, concretely, should we do here? My takeaway from 

> Any gap in support should be treated as a transient bug in TypeScript

is that we should preferably have TS "fix" their analysis so that both `if (oldBar != null && )` and `if (oldBar?.x === somethingNonNullish)` should result in `foo.baz = 1` being allowed by the type system. Assuming that's implemented, there would be nothing for us to do, since the autofix in the OP would no longer result in invalid TS code. Right?

**Comment by kirkwaiblinger** [comment]
Unless your view is that `if (oldBar && etc)` implies that `if (oldBar != null && etc)` should also work, but doesn't imply that `if (oldBar?.x === nonNullish)` should work?

And if that's the case... then what, concretely, would a PR do? Attempt to replicate the logic to detect this pattern so that we can avoid providing a linter report?

**Comment by JoshuaKGoldberg** [comment]
> is that we should preferably have TS "fix" their analysis 

That's what I was thinking, yeah. 

> Assuming that's implemented, there would be nothing for us to do, since the autofix in the OP would no longer result in invalid TS code. Right?

...I think you just helped me understand this more concretely. Whoa. Ok. I think you're right and this should be a wontfix as a TypeScript bug, not a typescript-eslint thing. Right?

**Comment by kirkwaiblinger** [comment]
> > is that we should preferably have TS "fix" their analysis
> 
> That's what I was thinking, yeah.
> 
> > Assuming that's implemented, there would be nothing for us to do, since the autofix in the OP would no longer result in invalid TS code. Right?
> 
> ...I think you just helped me understand this more concretely. Whoa. Ok. I think you're right and this should be a wontfix as a TypeScript bug, not a typescript-eslint thing. Right?

Progress! Yep, my thought is that we should reasonably expect TS to error in both the "before" and "after" cases (which was my expectation), or to error in neither the "before" nor "after" cases (which is more in line with your preference). In either of those worlds, there is nothing for us to do. So IMO we'd just consider this external

**Comment by JoshuaKGoldberg** [comment]
Agreed. Closing - but if anybody wants to post protests we have time before the stalebot kicks in.

This was such an interesting bug, thanks for filing @OpportunityLiu!
